validator/client/beacon-api/doppelganger.go

```diff
@@ -0,0 +1,229 @@
+	//   database for the last 2 epochs, we consider there is no doppelganger.
+	//   This is typically the case when we reboot the validator client.
+	// - If some validators we want to check doppelganger existence were NOT live
+	//   in local antislashing for the last two epochs, then we check onchain if there is
+	//   some liveness for these validators. If yes, we consider there is a doppelganger.
+
+	// Check inputs are correct
+	if in == nil || in.ValidatorRequests == nil || len(in.ValidatorRequests) == 0 {
+		return &ethpb.DoppelGangerResponse{
+			Responses: []*ethpb.DoppelGangerResponse_ValidatorResponse{},
+		}, nil
+	}
+
+	validatorRequests := in.ValidatorRequests
+
+	// Prepare response
+	stringPubKeys := make([]string, len(validatorRequests))
+	stringPubKeyToDoppelGangerHelper := make(map[string]DoppelGangerHelper, len(validatorRequests))
+
+	for i, vr := range validatorRequests {
+		if vr == nil {
+			return nil, errors.New("validator request is nil")
+		}
+
+		pubKey := vr.PublicKey
+		stringPubKey := hexutil.Encode(pubKey)
+		stringPubKeys[i] = stringPubKey
+
+		stringPubKeyToDoppelGangerHelper[stringPubKey] = DoppelGangerHelper{
+			validatorEpoch: vr.Epoch,
+			response: &ethpb.DoppelGangerResponse_ValidatorResponse{
+				PublicKey:       pubKey,
+				DuplicateExists: false,
+			},
+		}
+	}
+
+	// Retrieve fork version -- Return early if we are in phase0
+	forkResponse, err := c.getFork(ctx)
+	if err != nil || forkResponse == nil || forkResponse.Data == nil {
+		return nil, errors.Wrapf(err, "failed to get fork")
+	}
+
+	forkVersionBytes, err := hexutil.Decode(forkResponse.Data.CurrentVersion)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to decode fork version")
+	}
+
+	forkVersion := binary.LittleEndian.Uint32(forkVersionBytes)
+
+	if forkVersion == version.Phase0 {
+		log.Info("Skipping doppelganger check for Phase 0")
+		return buildResponse(stringPubKeys, stringPubKeyToDoppelGangerHelper), nil
+	}
+
+	// Retrieve current epoch
+	headers, err := c.getHeaders(ctx)
+	if err != nil || headers == nil || headers.Data == nil || len(headers.Data) == 0 ||
+		headers.Data[0].Header == nil || headers.Data[0].Header.Message == nil {
+		return nil, errors.Wrapf(err, "failed to get headers")
+	}
+
+	headSlotUint64, err := strconv.ParseUint(headers.Data[0].Header.Message.Slot, 10, 64)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse head slot")
+	}
+
+	headSlot := types.Slot(headSlotUint64)
+	currentEpoch := slots.ToEpoch(headSlot)
+
+	// Extract input pubkeys we did not validate for the 2 last epochs.
+	// If we detect onchain liveness for these keys during the 2 last epochs, it may
+	// exist a doppelganger somewhere.
+	var notRecentStringPubKeys []string
+
+	for _, spk := range stringPubKeys {
+		dph, ok := stringPubKeyToDoppelGangerHelper[spk]
+		if !ok {
+			return nil, errors.New("failed to retrieve doppelganger helper from string public key")
```

Comment:
```suggestion
			return nil, errors.New("failed to retrieve doppelganger info from string public key")
```
In case you rename `DoppelGangerHelper` to `DoppelGangerInfo`